### DISCUSSION TITLE
Should I expect anything from file previews or is it going to be just another file manager?

### DISCUSSION THREAD

**Discussion by txtyash** [kamiyaa/joshuto#199] [General]
* There's quite a few cli file managers:
        *   Joshuto
        *   lf
        *   ranger
        *   xplr
        *   nnn

Is this going to be yet another file manager rewrite in rust? I think the only feature that could set this file manager apart is file previews. If it has good file previews and especially a universal way of previewing images then that'd be great. Should I expect this file manager to have that?

Whenever I bring up this topic of image previews being important, some users just freak out and call it bloat, unnecessary, & useless. I literally get told to go use sxiv or an image viewer for that. They tell me that things can be configured and I should spend some time writing and maintaining configs rather than using a cli file manager with essential features right away. Is it so much fun to just keep configuring things? Or maybe it's just me who doesn't understand the point of creating similar cli file managers in multiple languages. If anyone could explain this to me then I'd appreciate it.

**Comment by txtyash** [comment]
I'd like to have something like the first preview gif at https://github.com/jarun/nnn with 0 config(although nnn requires configuration for that too). Ig everyone would like to have a file manager like that with 0 config, yet there's no project that wants to do this. This is the problem with most cli projects where users are bombarded with philosophy : (

**Comment by kamiyaa** [comment]
Here are my thoughts on this topic.

**Regarding image previews:**

Image previews are great in terms of user experience. No argument there.
And it sucks that terminal file manager require so much configurations to get image previews to work.

However, I think the fundamental problem here is there is no standard way of displaying images on the terminal.
Terminals were never designed with images in mind so all the solutions out there are very niche.
Either the image previews only work on a specific platform with a specific library (w3img works on X11, but fails on wayland) or it only works with select terminals (kitty, etc.)

So its not so much that these file managers don't want to support image previews, they're just really hard to get right and get it to work for everyone.
If there was a standard and elegant solution, I believe nearly all terminal file managers would implement it out of the box.

The only current solution (that I know of and while maintaining customizability and flexibility) is to have a script file that tries its best to adhere to everyone's setup.
Currently Joshuto's image preview support could be better, but I believe with time the script should get to the point where it works for 99.9% of users and just works out of the box (assuming the user copies the script).

The development of joshuto's clipboard support was similar. (https://github.com/kamiyaa/joshuto/blob/main/src/commands/file_ops.rs#L117)
Originally, it only supported a few clipboard managers on Linux.
But with time, Mac clipboard support was added by a user. So now joshuto works with Mac clipboard out of the box

**Regarding the goals of this project:**

The original main goals of this project was:
1. To facilitate me learning Rust (in which it succeeded spectacularly)
2. Make a `ranger` alternative that is faster and easier to configure (in my opinion)

I don't think there's any downsides to having multiple file managers to choose from.
All of them have perks and drawbacks and it's really up to the user to decide which they want to stick with.

Cheers!

**Reply by txtyash** [reply]
@kamiyaa I believe I've also seen some of your videos on web dev. Is it possible to have image previews inside the browser? I don't mean opening a new tab for every single image scrolled but update the same tab for every image scrolled and if possible it would be much better to have the preview of all images in that directory, in a bottom pane.

Assuming, every person uses a web browser ; )
